homeassistant/components/heos/services.py

```diff
@@ -71,3 +104,138 @@ async def _sign_out_handler(controller, service):
         await controller.sign_out()
     except HeosError as err:
         _LOGGER.error("Unable to sign out: %s", err)
+
+
+async def _join_handler(controller, hass, service):
```

Comment:
This is overly complex.  `create_group` and `update_group` are the same thing in the API (they both translate to `set_group` behind the scenes.  It's fine to call `update_group` directly and pass in the IDs directly after resolving.  We shouldn't be checking existing groups and merging.  If the user intends to update an existing group, they should supply the entire group definition, otherwise you break the ability to reduce members.

Reply:
Reducing members could also be done through the `unjoin()` service call, but you are right about the complexity. The new `HeosMediaPlayer.async_join()` method is quite simple now with the merging logic removed.